framework/src/utils/PetscSupport.C

```diff
@@ -518,8 +521,17 @@ petscSetDefaults(FEProblemBase & problem)
   {
     // dig out PETSc solver
     NonlinearSystemBase & nl = problem.getNonlinearSystemBase(nl_index);
-    PetscNonlinearSolver<Number> * petsc_solver =
-        dynamic_cast<PetscNonlinearSolver<Number> *>(nl.nonlinearSolver());
+    auto * const petsc_solver = cast_ptr<PetscNonlinearSolver<Number> *>(nl.nonlinearSolver());
+    auto * const sys_matrix = petsc_solver->system().request_matrix("System Matrix");
+    if (sys_matrix && problem.solverParams()._type != Moose::ST_JFNK)
```

Comment:
```suggestion
    // Prefix the name of the system matrix with the name of the system
    if (sys_matrix && problem.solverParams()._type != Moose::ST_JFNK)
```
so JFNK has a system matrix ? why the two conditions in the if

Reply:
It does have a system matrix, it just doesn't have it initialized, e.g no nonzeroes allocated for it. I have not heard anyone complain about this solution so I have never tried to explore a design in which we never create a `SparseMatrix` at all if we are doing JFNK. Perhaps others tried before me, I do not know